File: tic_tac_toe.py
import numpy

#board = numpy.array([["_"]*3]*3)

#for i in range(3):
#    for j in range(3):
#        print(board[i][j],end=" ")
#    print()

def ismoveleft(board):
    #global board
    for i in range(3):
        for j in range(3):
            if board[i][j]=="_":
                return True
    return False
def iswinner(board):
    for i in range(3):
        if board[i][0]==board[i][1]==board[i][2]=="X":
            return -10
        elif board[i][0]==board[i][1]==board[i][2]=="O":
            return 10
    
    #vertical
    for i in range(3):
        if board[0][i]==board[1][i]==board[2][i]=="X":
            return -10
        elif board[0][i]==board[1][i]==board[2][i]=="O":
            return 10
    
    #digonals
    if board[0][0]==board[1][1]==board[2][2]=="X":
        return -10
    if board[0][0]==board[1][1]==board[2][2]=="O":
        return 10
    if board[0][2]==board[1][1]==board[2][0]=="X":
        return -10
    if board[0][2]==board[1][1]==board[2][0]=="O":
        return 10
    
    return 0
    
def minmax(board,depth,ismx):
    result = iswinner(board)
    if result==10 or result ==-10:
        return result
    if ismoveleft(board)==False:
        return 0
    if ismx:
        bestscore =-10**9
        for i in range(3):
            for j in range(3):
                if board[i][j]=="_":
                    board[i][j]="O"
                    score =minmax(board,depth+1,False)
                    bestscore=max(score,bestscore)
                    board[i][j]="_"
        return bestscore
    else:
        bestscore =10**9
        for i in range(3):
            for j in range(3):
                if board[i][j]=="_":
                    board[i][j]="X"
                    score =minmax(board,depth+1,True)
                    bestscore=min(score,bestscore)
                    board[i][j]="_"
        return bestscore

def bestmove(board):
    bestscore=-10**9
    move=[-1,-1]        
    for i in range(3):
        for j in range(3):
            if board[i][j]=="_":
                board[i][j]="O"
                score = minmax(board,0,False)
                board[i][j]="_"
                if score>bestscore:
                    bestscore = score
                    move[0]=i
                    move[1]=j
                    #print("min",move)
    board[move[0]][move[1]]="O"

        
if __name__=="__main__":
    board = numpy.array([["_"]*3]*3)
    for i in range(9):
        human =int(input())
        if human<3:
            board[0][human]="X"
        elif 3<=human<6:
            board[1][human%3]="X"
        elif 6<=human<9:
            board[2][human%3]="X"
        for j in range(3):
            for k in range(3):
               print(board[j][k],end=" ")
            print()
        print()
        win = iswinner(board)
        if ismoveleft(board)==False and win==0:
            print("tie")
            break
        bestmove(board);    
        for j in range(3):
            for k in range(3):
                print(board[j][k],end=" ")
            print()
        if win==10:
            print("O wins")
            break
        elif win==-10:
            print("x wins")
            break
        

File: TicTacToe.c
#include<stdio.h>
#include<math.h>
char board[3][3];

// human move
int human_move(int pos)
{
        if(pos == 1 && board[0][0]=='_'){
            board[0][0]='X';
            return 1;
        }
        else if(pos == 2 && board[0][1]=='_'){
            board[0][1]='X';
            return 1;
        }
        else if(pos == 3 && board[0][2]=='_'){
            board[0][2]='X';
            return 1;
        }
        else if(pos == 4 && board[1][0]=='_'){
            board[1][0]='X';
            return 1;
        }
        else if(pos == 5 && board[1][1]=='_'){
            board[1][1]='X';
            return 1;
        }
        else if(pos == 6 && board[1][2]=='_'){
            board[1][2]='X';
            return 1;
        }
        else if(pos == 7 && board[2][0]=='_'){
            board[2][0]='X';
            return 1;
        }
        else if(pos == 8 && board[2][1]=='_'){
            board[2][1]='X';
            return 1;
        }
        else if(pos == 9 && board[2][2]=='_'){
            board[2][2]='X';
            return 1;
        }
        else{
            printf("Invalid position \n");
            return 0;
        }

}

void printCurrent(){
    int i,j;
    for(i=0;i<3;i++){
        for(j=0;j<3;j++){
            printf("%c ",board[i][j]);
        }
        printf("\n");
    }
}

int isWinner(){
    int i,j;
    //horizontal
    for(i=0;i<3;i++){
        if(board[i][0]=='X' && board[i][1]=='X' && board[i][2]=='X')        return -10;
        else if(board[i][0]=='O' && board[i][1]=='O' && board[i][2]=='O')        return 10;
    }
    //vertical
    for(i=0;i<3;i++){
        if(board[0][i]=='X' && board[1][i]=='X' && board[2][i]=='X')        return -10;
        else if(board[0][i]=='O' && board[1][i]=='O' && board[2][i]=='O')        return 10;
    }
    //diagonal
    if(board[0][0]=='X' && board[1][1]=='X' && board[2][2]=='X')        return -10;
    if(board[0][2]=='X' && board[1][1]=='X' && board[2][0]=='X')        return -10;
    if(board[0][0]=='O' && board[1][1]=='O' && board[2][2]=='O')        return  10;
    if(board[0][2]=='O' && board[1][1]=='O' && board[2][0]=='O')        return  10;
    // if no match found
    return 0;
}

int anyMoveLeft(){
    int i,j;
    for(i=0;i<3;i++){
        for(j=0;j<3;j++){
            if(board[i][j]=='_')    return 1;
        }
    }
    return 0;
}

int minmax(int depth,int currentmove){
    int res = isWinner();
    if(res==10) return 10;
    if(res==-10) return -10;
    if(!anyMoveLeft()) return 0;

    if(currentmove){
        int bestscore=-100000;
        int score;
        int i,j;
        for(i=0;i<3;i++){
            for(j=0;j<3;j++){
                if(board[i][j]=='_'){
                    board[i][j]='O';
                    score = minmax(depth+1,0);
                    board[i][j]='_';
                    if(score>=bestscore)
                        bestscore =score;
                }
            }
        }
        return bestscore;
    }
    else{
        int bestscore =100000;
        int score;
        int i,j;
        for(i=0;i<3;i++){
            for(j=0;j<3;j++){
                if(board[i][j]=='_'){
                    board[i][j]='X';
                    score = minmax(depth+1,1);
                    board[i][j]='_';
                    if(score<=bestscore)
                        bestscore =score;
                }
            }
        }
        return bestscore;
    }
}

void bestMove(){
    int i,j;
    int bestscore = -10000;
    int move[2];
    int score;
    move[0]=-1;
    move[1]=-1;

    for(i=0;i<3;i++){
        for(j=0;j<3;j++){
            if(board[i][j]=='_'){
                board[i][j]='O';
                score = minmax(0,0);
                board[i][j]='_';
                if(score>bestscore){
                    bestscore = score;
                    move[0]=i;
                    move[1]=j;
                }

            }
        }
    }

    board[move[0]][move[1]] = 'O';

}

void main()
{
    int i,j,win;
    //empty board
    for(i=0;i<3;i++){
        for(j=0;j<3;j++){
            board[i][j]='_';
        }
    }

    while(1){
        int pos,hmove;
        printf("your move \n");
        scanf("%d",&pos);
        hmove = human_move(pos);

        if (!hmove)
            continue;

        printCurrent();
        win = isWinner();

        if(!anyMoveLeft() && win==0){
            printf("TIE\n");
            break;
        }

        bestMove();
        win = isWinner();

        printf("computer moved \n");
        printCurrent();

        if(win ==10){
            printf("You lose\n");
            break;
        }
    }
}
